File: india_compliance/gst_india/overrides/item_tax_template.py
import frappe
from frappe import _
from frappe.utils import flt, rounded

from india_compliance.gst_india.overrides.transaction import get_valid_accounts
from india_compliance.gst_india.utils import get_gst_accounts_by_type


def validate(doc, method=None):
    doc.gst_rate = flt(doc.gst_rate)
    validate_zero_tax_options(doc)
    validate_tax_rates(doc)


def validate_zero_tax_options(doc):
    if doc.gst_treatment != "Taxable":
        doc.gst_rate = 0
        return

    if doc.gst_rate == 0:
        frappe.throw(
            _("GST Rate cannot be zero for <strong>Taxable</strong> GST Treatment"),
            title=_("Invalid GST Rate"),
        )


def validate_tax_rates(doc):
    if doc.gst_rate < 0 or doc.gst_rate > 100:
        frappe.throw(
            _("GST Rate should be between 0 and 100"), title=_("Invalid GST Rate")
        )

    if not doc.taxes:
        return

    __, intra_state_accounts, inter_state_accounts = get_valid_accounts(
        doc.company, for_sales=True, for_purchase=True, throw=False
    )

    if not intra_state_accounts and not inter_state_accounts:
        return

    invalid_tax_rates = {}
    for row in doc.taxes:
        tax_rate = abs(row.get("tax_rate") or 0)

        # check intra state
        if row.tax_type in intra_state_accounts and doc.gst_rate != tax_rate * 2:
            invalid_tax_rates[row.idx] = doc.gst_rate / 2

        # check inter state
        elif row.tax_type in inter_state_accounts and doc.gst_rate != tax_rate:
            invalid_tax_rates[row.idx] = doc.gst_rate

    if not invalid_tax_rates:
        return

    # throw
    message = (
        "Plese make sure account tax rates are in sync with GST rate mentioned."
        " Following rows have inconsistant tax rates: <br><br>"
    )

    for idx, tax_rate in invalid_tax_rates.items():
        message += f"Row #{idx} - should be {rounded(tax_rate, 2)}% <br>"

    frappe.throw(_(message), title=_("Invalid Tax Rates"))


@frappe.whitelist()
def get_valid_gst_accounts(company):
    frappe.has_permission("Item Tax Template", "read", throw=True)

    return [
        *get_valid_accounts(company, for_sales=True, for_purchase=True, throw=False),
        get_accounts_with_negative_rate(company),
    ]


def get_accounts_with_negative_rate(company):
    """
    - Include all Sales RCM accounts
    - Include all Purchase RCM accounts based on how taxes and charges template is created
    """
    negative_rate_accounts = list(
        get_gst_accounts_by_type(company, "Sales Reverse Charge", throw=False).values()
    )
    negative_rate_accounts += list(
        get_gst_accounts_by_type(company, "Output Refund", throw=False).values()
    )  # add refund accounts

    purchase_rcm_accounts = list(
        get_gst_accounts_by_type(
            company, "Purchase Reverse Charge", throw=False
        ).values()
    )

    if not purchase_rcm_accounts:
        return negative_rate_accounts

    add_deduct_tax = frappe.get_value(
        "Purchase Taxes and Charges",
        {
            "parenttype": "Purchase Taxes and Charges Template",
            "account_head": ["in", purchase_rcm_accounts],
        },
        "add_deduct_tax",
    )

    if add_deduct_tax == "Add":
        negative_rate_accounts += purchase_rcm_accounts

    return negative_rate_accounts

File: india_compliance/gst_india/client_scripts/item_tax_template.js
frappe.ui.form.on("Item Tax Template", {
    refresh: show_missing_accounts_banner,
    fetch_gst_accounts: fetch_and_update_missing_gst_accounts,
    async gst_rate(frm) {
        if (frm.doc.gst_rate === null) return;

        await Promise.all(
            frm.doc.taxes.map(async row => {
                const tax_rate = await get_tax_rate_for_account(frm, row.tax_type);
                if (tax_rate == null) return;

                row.tax_rate = tax_rate;
            })
        );

        frm.refresh_field("taxes");
    },
});

async function show_missing_accounts_banner(frm) {
    if (frm.doc.gst_treatment === "Non-GST" || frm.doc.__islocal) return;

    const missing_accounts = await get_missing_gst_accounts(frm);
    if (!missing_accounts) return;

    // show banner
    frm.dashboard.add_comment(
        __(`<strong>Missing GST Accounts:</strong> {0}`, [missing_accounts.join(", ")]),
        "orange",
        true
    );
}

async function fetch_and_update_missing_gst_accounts(frm) {
    const missing_accounts = await get_missing_gst_accounts(frm);
    if (!missing_accounts) return;

    // cleanup existing empty rows
    frm.doc.taxes = frm.doc.taxes.filter(row => row.tax_type);

    // add missing rows
    await Promise.all(
        missing_accounts.map(async account => {
            const tax_rate = await get_tax_rate_for_account(frm, account);
            frm.add_child("taxes", { tax_type: account, tax_rate: tax_rate });
        })
    );

    frm.refresh_field("taxes");
}

async function get_tax_rate_for_account(frm, account) {
    const gst_rate = frm.doc.gst_rate;
    if (!gst_rate) return 0;

    const gst_accounts = await get_gst_accounts(frm);
    if (!gst_accounts) return null;

    const [_, intra_state_accounts, inter_state_accounts] = gst_accounts;

    if (intra_state_accounts.includes(account)) return gst_rate / 2;
    else if (inter_state_accounts.includes(account)) return gst_rate;
    else return null;
}

async function get_missing_gst_accounts(frm) {
    let gst_accounts = await get_gst_accounts(frm);
    if (!gst_accounts) return;

    const all_gst_accounts = gst_accounts[0];
    const template_accounts = frm.doc.taxes.map(t => t.tax_type);
    const missing_accounts = all_gst_accounts.filter(
        a => a && !template_accounts.includes(a)
    );

    if (missing_accounts.length) return missing_accounts;
}

async function get_gst_accounts(frm) {
    const company = frm.doc.company;

    // cache company gst accounts
    if (!frm._company_gst_accounts?.[company]) {
        frm._company_gst_accounts = frm._company_gst_accounts || {};
        const { message } = await frappe.call({
            method: "india_compliance.gst_india.overrides.transaction.get_valid_gst_accounts",
            args: { company: company },
        });

        frm._company_gst_accounts[company] = message;
    }

    if (!frm._company_gst_accounts[company]) return;
    return frm._company_gst_accounts[company];
}
